tests/upgrades/test_virtwho.py

```diff
@@ -0,0 +1,195 @@
+            u'hypervisor_server': self.hypervisor_server,
+            u'organization_id': self.org.id,
+            u'filtering_mode': 'none',
+            u'satellite_url': self.satellite_url,
+        }
+        if self.hypervisor_type == 'libvirt':
+            args[u'hypervisor_username'] = self.hypervisor_username
+        elif self.hypervisor_type == 'kubevirt':
+            args[u'kubeconfig'] = self.hypervisor_config_file
+        else:
+            args[u'hypervisor_username'] = self.hypervisor_username
+            args[u'hypervisor_password'] = self.hypervisor_password
+        return args
+
+    def _try_to_get_guest_bonus(self, hypervisor_name, sku):
+        subscriptions = entities.Subscription().search(
+            query={'search': '{0}'.format(sku)})
+        for item in subscriptions:
+            item = item.read_json()
+            if hypervisor_name.lower() in item['hypervisor']['name']:
+                return item['id']
+
+    def _get_guest_bonus(self, hypervisor_name, sku):
+        vdc_id, time = wait_for(self._try_to_get_guest_bonus,
+                                func_args=(hypervisor_name, sku),
+                                fail_condition=None,
+                                timeout=15,
+                                delay=1)
+        return vdc_id
+
+    @pre_upgrade
+    def test_pre_create_virt_who_configuration(self):
+        """Create and deploy virt-who configuration
+
+        :id: preupgrade-a36cbe89-47a2-422f-9881-0f86bea0e24e
+
+        :steps: In Preupgrade Satellite, Create and deploy virt-who configuration.
+
+        :expectedresults:
+            1. Config can be created and deployed by command.
+            2. No error msg in /var/log/rhsm/rhsm.log.
+            3. Report is sent to satellite.
+            4. Virtual sku can be generated and attached.
+        """
+        args = self._make_virtwho_configure()
+        args.update({'name': self.name})
+        vhd = entities.VirtWhoConfig(**args).create()
+        self.assertEqual(vhd.status, 'unknown')
+        command = get_configure_command(vhd.id)
+        hypervisor_name, guest_name = deploy_configure_by_command(
+            command, debug=True)
+        self.assertEqual(
+            entities.VirtWhoConfig().search(
+                query={'search': 'name={}'.format(self.name)})[0].status,
+            'ok')
+        hosts = [
+            (hypervisor_name, 'product_id={} and type=NORMAL'.format(
+                self.vdc_physical)),
+            (guest_name, 'product_id={} and type=STACK_DERIVED'.format(
+                self.vdc_physical))]
+        for hostname, sku in hosts:
+            if 'type=NORMAL' in sku:
+                subscriptions = entities.Subscription().search(
+                    query={'search': '{0}'.format(sku)})
+                vdc_id = subscriptions[0].id
+            if 'type=STACK_DERIVED' in sku:
+                vdc_id = self._get_guest_bonus(hypervisor_name, sku)
```

Comment:
I'm not a huge fan of this approach, where we create list of tuples and then, in the loop, the very first thing is to assign new variable based on one of tuple values. I would rather create variables before loop and use them directly:
```python
normal_vdc_id = entities.Subscription().search(query={'search': 'product_id={} and type=NORMAL'.format(self.vdc_physical)})
stacked_vdc_id = self._get_guest_bonus(hypervisor_name, 'product_id={} and type=STACK_DERIVED'.format(self.vdc_physical))
hosts = [
    (hypervisor_name, normal_vdc_id),
    (guest_name, stacked_vdc_id)
]
for hostname, vdc_id in hosts:
    host, time = wait_for(entities.Host().search, ...)
```
(of course with better formatting than above)

Reply:
I agree with you, thanks for the better way

Reply:
I forgot this, sorry, the vdc_id of the guest only can be searched after the hypervisor attached. That is why we designed here. So, maybe we should leave this part of code as it is.:)

Reply:
I see. I have one more idea, but it's not clear improvement over what we have here, so let's leave it as is - unless someone else jumps in with some other good approach.